src/KafkaW/Consumer.cpp

```diff
@@ -127,17 +133,36 @@ bool Consumer::topicPresent(const std::string &TopicName) {
   return true;
 }
 
+/// Updates the RdKafka Metadata pointer. If broker is unavailable, keeps trying
+/// to connect every 500ms, logging messages every few seconds.
 void Consumer::updateMetadata() {
+  short LoopCounter = 0;
+  while (!metadataCall()) {
+    if (LoopCounter == 7) {
+      Logger->error("Cannot contact broker, retrying until connection is "
+                    "established...");
+      LoopCounter = 0;
+    }
+    LoopCounter++;
+  }
+  Logger->info("Connection with broker established.");
+}
+
+bool Consumer::metadataCall() {
   RdKafka::Metadata *MetadataPtr = nullptr;
   auto ErrorCode = KafkaConsumer->metadata(
       true, nullptr, &MetadataPtr, ConsumerBrokerSettings.MetadataTimeoutMS);
```

Comment:
I have had problems before caused by metadata calls taking too long and timing out before getting a response. Even with a timeout of 1 second. Solving this problem in a satisfactory way is hard as the call to `metadata()` is a blocking call. A `maybe` acceptable work around for now could be to execute that call in a separate thread (with a timeout of let's say 10 seconds) so as to not block the current thread.

As a bare minimum, you should add the following Doxygen comment `/// /todo Re-design function to execute blocking call in separate thread.`.

Reply:
Reading the code a bit further, it looks like you actually are executing the code in a separate thread already, correct? Regardless, the code still needs some changes to nicely deal with longer timeouts.

Reply:
Yes, each consumer is async launched as a separate thread.
We recently also discovered metadata call timing out on kafkacow running on windows. We're not sure what causes it. We'll see if that's related.
We will discuss it with @matthew-d-jones .

Reply:
I have created a JIRA ticket for this issue: https://jira.esss.lu.se/browse/DM-1599
I think this temporarily resolves the issue for but I also think we need to fix the problem as soon as possible.